### PR TITLE
feat: add human-in-the-loop tool approval system

### PR DESCRIPTION
**Pull request by MaderHatt3r**
Add a comprehensive tool approval system that requires explicit user consent before AI agents execute tools, with granular control at both the parent tool and child function level.

Backend:
- Add ENABLE_TOOL_APPROVAL config setting (env var + admin toggle)
- Add ToolApprovalManager with in-memory state for pending approvals, always-approved permissions, and YOLO mode tracking
- Integrate approval checks into middleware for both streaming and non-streaming tool calls, with per-function granularity
- Add Socket.IO handlers for approval responses, always-approved CRUD, and YOLO mode management
- Add `BUILTIN_TOOL_CATEGORIES` canonical mapping and `GET /tools/builtin/categories` endpoint for dynamic builtin tool category metadata

Frontend:
- Add ToolApprovalModal with Allow Once, Always Approve (per-function and per-tool), and Decline actions
- Add YOLO Mode and Always Allowed sub-menus to IntegrationsMenu with per-tool toggles and tree-view revocation UI
- Add toolApproval Svelte store for socket-based state management
- Add cross-tab coordination via BroadcastChannel
- Show inline warning for unrecognized tool calls in approval modal

**Contribution Note**:

This feature was initially developed by @Barbaronno and extended with granular per-function/per-tool permissions, YOLO mode, an always-allowed revocation UI, and unrecognized tool call warnings.

This submitted commit adds @Barbaronno as a co-author.

See the original PR at: https://github.com/open-webui/open-webui/pull/16913

# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) to discuss your idea/fix with the community before creating a pull request, and describe your changes before submitting a pull request.

This is to ensure large feature PRs are discussed with the community first, before starting work on it. If the community does not want this feature or it is not relevant for Open WebUI as a project, it can be identified in the discussion before working on the feature and submitting the PR.

#### First-time Contributor Acknowledgement

See the discussion at: https://github.com/open-webui/open-webui/discussions/16701

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Verify that the pull request targets the `dev` branch. **PRs targeting `main` will be immediately closed.**
- [x] **Description:** Provide a concise description of the changes made in this pull request down below.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Add docs in [Open WebUI Docs Repository](https://github.com/open-webui/docs). Document user-facing behavior, environment variables, public APIs/interfaces, or deployment steps.
  - [x] Documentation PR: https://github.com/open-webui/docs/pull/1095
- [x] **Dependencies:** Are there any new or upgraded dependencies? If so, explain why, update the changelog/docs, and include any compatibility notes. Actually run the code/function that uses updated library to ensure it doesn't crash.
- [x] **Testing:** Perform manual tests to **verify the implemented fix/feature works as intended AND does not break any other functionality**. Include reproducible steps to demonstrate the issue before the fix. Test edge cases (URL encoding, HTML entities, types). Take this as an opportunity to **make screenshots of the feature/fix and include them in the PR description**.
- [x] **Agentic AI Code:** Confirm this Pull Request is **not written by any AI Agent** or has at least **gone through additional human review AND manual testing**. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
  - [x] Both the description and code were heavily helped by cursor with Opus 4.6 as my favored model, but I am responsible for reviewing and testing it. It made a fair number of odd decisions along the way that I corrected. For instance the backend storage data structure was a mess, but it cleaned things up nicely when I recommended the nested dictionary and tree-view structures.
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Design & Architecture:** Prefer smart defaults over adding new settings; use local state for ephemeral UI logic. Open a Discussion for major architectural or UX changes.
- [x] **Git Hygiene:** Keep PRs atomic (one logical change). Clean up commits and rebase on `dev` to ensure no unrelated commits (e.g. from `main`) are included. Push updates to the existing PR branch instead of closing and reopening.
  - [x] **Refactor Notice:** Minor refactor included. built-in tools for models in the admin panel were hard-coded before. Moved this to a dictionary with an endpoint for the backend so they can be shared with the new features. I kept this in a separate commit in case you'd like a separate PR.
- [x] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

Adds a human-in-the-loop (HITL) tool approval system that requires explicit user consent before AI agents execute tools during chat. When enabled, every tool call pauses and presents a modal asking the user to **Allow Once**, **Always Approve**, or **Decline**. Approval granularity operates at both the individual function level (e.g. `read_file`) and the parent tool level (e.g. `mcp:desktop_commander`). A **YOLO Mode** option is available for users who want to auto-approve all tool calls without prompts. All approval state is per-chat and in-memory — nothing is persisted to disk or database.

### Added

- `ENABLE_TOOL_APPROVAL` configuration setting — toggleable via environment variable or Admin Settings > General. Defaults to `true`.
- `ToolApprovalManager` backend class (`backend/open_webui/utils/tool_approval.py`) managing pending approvals, always-approved permissions, and YOLO mode state, all in-memory and per-chat.
- Middleware integration for both streaming and non-streaming tool calls that checks approval status before execution, with per-function granularity.
- Socket.IO handlers (`backend/open_webui/socket/main.py`) for real-time approval responses, always-approved CRUD, YOLO mode management, and pending approval recovery.
- `ToolApprovalModal` frontend component with four actions: Allow Once, Always Approve (per-function), Always Approve (per-tool), and Decline.
- YOLO Mode sub-menu in IntegrationsMenu with a master "YOLO All Tools" toggle and per-category toggles organized into Default Features (gated by model capabilities), Builtin Tools (loaded dynamically from backend), and Workspace Tools sections.
- Always Allowed sub-menu in IntegrationsMenu with a tree-view display of approved parent tools and their child functions, with individual and bulk revocation.
- `toolApproval` Svelte store (`src/lib/stores/toolApproval.ts`) for socket-based frontend state management.
- Cross-tab coordination via `BroadcastChannel` (`src/lib/utils/toolApprovalChannel.ts`) to prevent duplicate approval modals across browser tabs.
- TypeScript interfaces for tool approval data structures (`src/lib/types/index.ts`).
- Inline warning banner in approval modal when a tool call has an unrecognized parent tool ID.
- `BUILTIN_TOOL_CATEGORIES` dict in `tools.py` as single source of truth for builtin tool category names, descriptions, and function mappings.
- `GET /tools/builtin/categories` API endpoint serving builtin tool metadata to authenticated users.

### Changed

- **Refactor Notice:** built-in tools for models in the admin panel were hard-coded before. Moved this to a dictionary with an endpoint for the backend.
  - `BuiltinTools.svelte` admin component fetches categories from backend API instead of hardcoding.
- `IntegrationsMenu` now accepts a `chatId` prop and displays YOLO Mode and Always Allowed entries when a chat is active.
- `MessageInput` passes `chatId` down to `IntegrationsMenu`.
- `Chat.svelte` passes `chatId` to the message input component.
- `+layout.svelte` handles `tool:approval_request` and `tool:approval_status` socket events globally.
- Admin config endpoints (`routers/auths.py`) expose and accept `ENABLE_TOOL_APPROVAL`.
- MCP tools and direct server tools in middleware now carry a `tool_id` field for parent tool resolution.

### Deprecated

- N/A

### Removed

- N/A

### Fixed

- N/A

### Security

- N/A

### Breaking Changes

- N/A

---

### Additional Information

- **Contribution Note**: This feature was initially developed by @Barbaronno and extended with granular per-function/per-tool permissions, YOLO mode, an always-allowed revocation UI, and unrecognized tool call warnings.
  - This submitted commit adds @Barbaronno as a co-author.
  - See the original PR at: https://github.com/open-webui/open-webui/pull/16913
- **Documentation PR**: https://github.com/open-webui/docs/pull/1095
- All approval state lives in-memory on the backend (`ToolApprovalManager` singleton) and resets on server restart. No database migrations or schema changes.
- The `ENABLE_TOOL_APPROVAL` env var can be set to `false` to disable the feature entirely, bypassing all approval checks.
- Approval granularity: functions are identified by their callable name (e.g. `read_file`), parent tools by their registered ID (e.g. `mcp:desktop_commander`, `direct_server:http://...`).
- YOLO Mode and Always Approved are scoped per-chat, so enabling YOLO in one conversation does not affect others.

### Screenshots or Videos

**Scenario 1.a** — YOLO a single tool in a new chat
- YOLO "Web Search" in a new chat, then 
- Prompt: `Find documentation for aws strands.`
**Result:** No permission was asked to use web search tool.
image

**Scenario 1.b** — Disable YOLO in the same chat
- In the same chat, turn YOLO off for "Web Search". 
- Prompt: `Great! Now find documentation for dynamodb.`
**Result:** Approval modal prompts for `search_web` / `builtin:search_web`.
image

**Scenario 1.c** —  Always allow function
- Click always allow for the function
**Result:** Search completed and function added to always allow list.
image

**Scenario 1.d** — Always allow tool
- Prompt: `Great! Now generate an image of a cat riding a donkey.`
- Click always allow at tool level
**Result:** Result: Always allow menu now shows the always allowed function as well as the tool. `All functions approved` displayed for tool-level approval
image

**Scenario 1.e** — Revoke function access
- Click the red x to revoke access for a web search
- Prompt: Now search for documentation for vLLM
**Result:** Modal prompt for allowing a web search.
image

**Scenario 1.f** — Cross-tab coordination
- Open the existing chat from Scenario 1.2 in a new browser window. How about my phone!
- **Result:** Approval modal prompts in the new device for web search.
Screenshot_20260220-200859

Note: These aren't the only test I preformed, but it became tedious to document it all in this format. 

I have tested allow once.

I have tested decline. It suggests to the tool that the user declined it. I have seen this redirect the llm mid stream. For instance it tried to access the knowledge base and I declined it and it did a web search instead.

image

I have tested with several functions under one tool and confirmed that the allow always ones allow, but not the individual functions for the same tool that aren't in the always allow list. 

I have tested it with the subagents tool. Note, this isn't supported at this time. It seemed to ignore permissions.

Error found during testing: Sometimes opening the menu to observe a recently changed setting wouldn't show the change until closing the menu and opening again. Fixed by awaiting the async operation. Updated existing feature commit and pushed. Re-tested.

Bonus screenshot:
Screenshot_20260216-172805~2

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

> [!NOTE]
> Deleting the CLA section will lead to immediate closure of your PR and it will not be merged in.

### CODE REVIEW

File: backend/open_webui/routers/tools.py [hunk 1]
```diff
@@ -50,6 +50,16 @@ def get_tool_module(request, tool_id, load_from_db=True):
     return tool_module
 
 
+############################
```
Comment (MaderHatt3r) [inline backend/open_webui/routers/tools.py:53]: Gets all built-in tools including Default Features and Builtin Tools. Dictionary contains a boolean flag stating whether it's a default feature or builtin tool. Used in admin panel for models and in yolo mode dialogue. Not needed in always allow dialogue since that is populated on-the-fly as tools are approved.  

i18n translations still applied as `{$i18n.t(categories[tool]?.name ?? tool)}`

File: src/lib/components/chat/MessageInput/IntegrationsMenu.svelte [hunk 2]
```diff
@@ -174,12 +370,12 @@
 												{/if}
 											</div>
 
-											<div class=" truncate">{filter?.name}</div>
```
Comment (MaderHatt3r) [inline src/lib/components/chat/MessageInput/IntegrationsMenu.svelte:177]: Looks like someone didn't run linting. Some of this file is linting changes only.

File: src/lib/components/chat/MessageInput/IntegrationsMenu.svelte [hunk 3]
```diff
@@ -392,12 +581,247 @@
...
+										<div class="shrink-0">
+											{#if category.id === 'web_search'}
+												<GlobeAlt />
+											{:else if category.id === 'image_generation'}
+												<Photo className="size-4" strokeWidth="1.5" />
+											{:else if category.id === 'code_interpreter'}
+												<Terminal className="size-3.5" strokeWidth="1.75" />
+											{/if}
+										</div>
+										<div class="truncate">{$i18n.t(category.name)}</div>
```
Comment (MaderHatt3r) [inline src/lib/components/chat/MessageInput/IntegrationsMenu.svelte]: Double checked that i18n is applied to the category has i18n translations.

Rename terrible variable name "category" and all related variables called category. These should be builtin tools and feature tools.

File: src/lib/components/chat/MessageInput/IntegrationsMenu.svelte [hunk 4]
```diff
@@ -392,12 +581,247 @@
...
+						<button
+							class="flex w-full justify-between gap-2 items-center px-3 py-1.5 text-sm cursor-pointer rounded-xl hover:bg-gray-50 dark:hover:bg-gray-800/50"
+							on:click={() => toggleToolYolo(toolId)}
+						>
+							<div class="flex-1 truncate">
+								<div class="flex flex-1 gap-2 items-center">
+									<div class="shrink-0">
+										<Wrench />
+									</div>
+									<div class="truncate">{tools[toolId].name}</div>
```
Comment (MaderHatt3r) [inline src/lib/components/chat/MessageInput/IntegrationsMenu.svelte]: User-defined tools don't need i18n.

File: src/lib/components/chat/MessageInput/IntegrationsMenu.svelte [hunk 5]
```diff
@@ -392,12 +581,247 @@
...
+										<div class="shrink-0">
+											{#if featureTool.id === 'web_search'}
+												<GlobeAlt />
+											{:else if featureTool.id === 'image_generation'}
+												<Photo className="size-4" strokeWidth="1.5" />
+											{:else if featureTool.id === 'code_interpreter'}
+												<Terminal className="size-3.5" strokeWidth="1.75" />
+											{/if}
+										</div>
+										<div class="truncate">{$i18n.t(featureTool.name)}</div>
```
Comment (MaderHatt3r) [inline src/lib/components/chat/MessageInput/IntegrationsMenu.svelte]: Feature and built-in tools get i18n like the admin panel.

**Review by MaderHatt3r — COMMENTED** [review]
Testing and review complete. Submitting for approval. Author confirms readiness to merge.

**Comment by MaderHatt3r** [comment]
I saw that this was part of the PR pipeline/ workflow, so I ran formatters and translations. Translations seem like they could detract from reviewing a PR though. Let me know if I should revert translations for a simpler PR.

```yaml
- name: Format Frontend
  run: npm run format

- name: Run i18next
  run: npm run i18n:parse

- name: Check for Changes After Format
  run: git diff --exit-code    # <-- fails if ANY file changed
  ```

**Comment by Classic298** [comment]
You reviewed your own PR? You should review your code before pushing it to GitHub, instead of pushing it on GitHub first, creating a PR, and only then reviewing your own code.

**Comment by Classic298** [comment]
Closing this PR.

The core feature, human-in-the-loop tool approval, is something Open WebUI needs (in my opinion), and Barbaronno's original work on #16913 is appreciated and still open. Picking up that work, rebasing it, and submitting a clean updated version would have been a welcome contribution - and Barbaronno even agreed to let you do that.

But that's not what happened here. [Your comment on #16913](https://github.com/open-webui/open-webui/pull/16913#issuecomment-2868432043) said you'd "get a fork created soon and a new pull request and link it to the old to get it closed". What arrived isn't a rebased version of the original feature, it's the original feature buried under significant undiscussed additions: "YOLO mode", per-function/per-tool granular permissions, tree-view revocation UI, a builtin tool categories refactor, and cross-tab BroadcastChannel coordination. **None of these went through a Discussion first**. The result is a PR that is larger and harder to review than the original, not easier.

There already is [the original discussion](https://github.com/open-webui/open-webui/discussions/16701), but these additional features were never discussed there or anywhere else. Are we sure they're wanted or needed?

Some specific feedback:

- "YOLO mode" is probably not an appropriate name for this project. Did you discuss and propose the name anywhere first?
- Granular per-tool permissions for builtin tools (toggling memory, time, channels, etc. individually) is an interesting idea, but it should be its own Discussion and its own PR, not bundled in here. And it needs discussion first, because adding up to 7 new toggles for builtin tool categories is UI bloat for something 99% of users will never need. Why would someone disable the time calculation tools? A case could be made for channels or notes, but not for every category.
- Review your own code before submitting, not after. Self-review comments posted on your own PR in the GitHub UI suggest the review step happened in the wrong order. The code should be reviewed locally before it's pushed, not annotated after the fact.

There are also multiple code quality issues:

1. no authorization on socket handlers: All the socket handlers in socket/main.py (tool:set_yolo, tool:add_always_approved, tool:clear_always_approved, etc.) only check that the session exists. They don't verify that the user owns the chat_id they're operating on. **Any authenticated user could potentially approve tools or set YOLO mode on someone else's chat**. This is a security vulnerability.
2. middleware.py: lines 4804-4809 are after return await response_handler(response, events) on line 4801. **This is unreachable, dead code.**
3. approval_manager.wait_for_approval() calls await event.wait() with no timeout. If a user closes their browser, navigates away, or loses connection, that coroutine hangs forever. **There's no cleanup mechanism.** The BroadcastChannel coordinator even acknowledges this in a comment: "Could take over, but we'll let it timeout per requirements" - but there is no timeout.
4. The ToolApprovalManager is a module-level singleton. **In any multi-worker setup (gunicorn, etc.), each worker has its own instance. The approval request could land on worker A while the approval response lands on worker B. The feature silently fails in production.**
5. The approval check block (~80 lines of approval gathering, event emission, waiting, and result handling) is copy-pasted nearly identically in two places in middleware.py, once for the non-streaming tool call path and once for the streaming path. One uses asyncio.gather for parallel approval, the other uses a sequential for loop. This inconsistency would cause different behavior depending on whether the model streams or not.
6. ToolApprovalRequest in types/index.ts has risk_level: 'high' | 'medium' | 'low' - this field is never set by the backend. It doesn't exist anywhere else in the codebase. **Dead code** shipped as a type definition.
7. ToolApprovalRequest in types/index.ts has risk_level but no parent_tool_id. The same interface in stores/toolApproval.ts has parent_tool_id but no risk_level. ToolApprovalData in types/index.ts has session_id, the store version doesn't.
8. The lightning bolt (YOLO), shield (Always Allowed), and X (revocation) icons are inline SVG blocks instead of using the project's established icon component pattern. Inconsistent with the rest of the codebase.
9. The bottom of toolApproval.ts has if (typeof window !== 'undefined') { import('$lib/stores').then(...) } - socket listeners registered as a module import side effect. Fragile with SSR, HMR, and testing.

Combined with #21685 and #21686 (both closed), this is three large PRs submitted in a single day, all probably AI-generated, all with scope and guideline issues. **Please slow down, read the contribution guidelines, go through the PR checklist slowly and ensure you actually follow the rules and open Discussions before writing code for new features.** The process exists to save everyone's time - yours and ours. (Keep PRs atomic (one logical change), Open a Discussion for major architectural or UX changes, ...)

Feel free to open discussions for your other features you added here. They would be great standalone PRs, but adding them on top of an already time-consuming PR doesn't help get the HITL reviewed any faster.

**Comment by MaderHatt3r** [comment]
This again was a feature I wrote from myself with the start from the original PR that I thought I would provide back. The additions seemed like welcome changes to the original author and those in the discussions. 

The original PR had some issues that I fixed and the main missing feature for me was Always Allow at a function level rather than the parent tool level. I started with mobile support for the ui, then made it granular. You say you'd be fine with a rebase of the original code, but it had a blanket disallow for all built-in tools without a way to "YOLO" them which seemed heavy handled, so I added YOLO support for blanket approval on certain tools. I was thinking a future addition could be a default YOLO for some of the built in tools (but not all of them. It kind of drives me crazy when you ask to "search" and the model thinks it needs to look up the knowledgebase and you can't stop it and suggest otherwise).

I'll admit, this was the first feature I worked on out of anything and I wasn't sure that I was going to continue to use this software because it seemed to be lacking and the original PR being completely abandoned wasn't a good sign for me. I spent less time than I probably should observing existing code for this reason hence missing the socket handler, existing svgs, etc. I originally wasn't going to put that much effort into this tool and ended up refactoring a lot of it once I decided to keep it, hence the dead code from direction changes etc.

I can clean up the code based on your code quality suggestions, but it sounds like you would still close it if even if I address the issues. Is this correct?

**Comment by MaderHatt3r** [comment]
Also, as far as the name `YOLO` is concerned, I'm sure the original authors intended it as an easter-egg-like joke, but it kind of became an industry standard in a sense, so feel free to suggest a different name,  but it is fairly accepted and referenced now.

----

In the current AI landscape, the term **"YOLO mode"** exists in two distinct worlds: **Agentic AI** (where it means "autonomous execution without approval") and **Computer Vision** (where it refers to the "You Only Look Once" real-time detection architecture). 

1. Agentic AI & Coding Tools

In these systems, "YOLO mode" is a specific setting that allows the AI to execute terminal commands, edit files, and call APIs without waiting for human confirmation.

- **Cursor:** An AI-powered code editor where "YOLO mode" allows the **Composer** feature to autonomously run terminal commands and apply code changes across multiple files to fix bugs or build features.
- **[Claude Code](https://code.claude.com/docs/en/overview):** Anthropic's command-line tool includes a mode (activated via `--dangerously-skip-permissions`) that users and documentation frequently refer to as "YOLO mode". It bypasses all permission prompts for file and system access.
- **[Traycer AI](https://docs.traycer.ai/tasks/yolo-mode):** Features a dedicated **"YOLO Mode"** button for its agentic workflows. It offers two sub-modes:
  - **Smart YOLO:** An intelligent orchestrator that adapts plans and specs in real-time as it executes.
  - **YOLO for Phases:** A more rigid automation based on predefined user configurations.
- **[Cline](https://docs.cline.bot/features/yolo-mode) (formerly Claude Dev):** An open-source VS Code extension that includes a "YOLO mode" checkbox to auto-approve all agent actions, from terminal execution to browser automation.
- **Roo Code:** A fork of Cline that maintains the "YOLO mode" feature for unattended autonomous coding.
- **[Sheet0](https://skywork.ai/blog/ai-agent/sheet0-yolo-mode/):** An AI agent for data extraction that uses "YOLO Mode" to scrape and process data from multiple sources in parallel without manual intervention.
- Computer Vision Frameworks (The Origin)

If you are looking for the **technical models** that pioneered the name, "YOLO" refers to a family of object detection models. These are typically managed by: 

- **[Ultralytics](https://docs.ultralytics.com/):** The primary maintainer of modern YOLO versions, including **YOLOv5, YOLOv8, YOLO11, and YOLO26**.
- **Academic/Corporate Variants:** Includes versions like **YOLOv10** (Tsinghua University), **PP-YOLO** (Baidu), and **YOLO-NAS** (Deci AI).
- Emergent Slang in AI Research

- **"YOLO Run":** A term used by researchers (like Jason Wei at OpenAI) to describe a training run where an ambitious model is implemented with hyperparameter settings based on intuition rather than extensive small-scale testing.

You can start by **enabling YOLO mode** in a tool like **Cursor** or **Cline** to see how much faster an autonomous agent can handle your repetitive coding tasks.

**Comment by Classic298** [comment]
> I can clean up the code based on your code quality suggestions, but it sounds like you would still close it if even if I address the issues. Is this correct?

Absolutely not! I would not close it under these assumptions

If you can copy the original PR, fix all the issues i commented under the original PR here: https://github.com/open-webui/open-webui/pull/16913#issuecomment-3939185585 (in case you missed it) then that would be greatly appreciated.

> I'll admit, this was the first feature I worked on out of anything and I wasn't sure that I was going to continue to use this software because it seemed to be lacking and the original PR being completely abandoned wasn't a good sign for me.

I agree. That's why I did a code review on the original PR to push things forward.

---

> Also, as far as the name YOLO is concerned, I'm sure the original authors intended it as an easter-egg-like joke

What?
Then who suggested YOLO?

**Comment by MaderHatt3r** [comment]
> You reviewed your own PR? You should review your code before pushing it to GitHub, instead of pushing it on GitHub first, creating a PR, and only then reviewing your own code.

The self-review comments are key decision explanations to try to help reviews go more smoothly, but I can completely exclude any and all comments in the future. The one change I made based on my own comment was while in draft mode which seems completely reasonable to me. The only other comment was a dangling question about your broken workflow.

**Comment by MaderHatt3r** [comment]
> > I can clean up the code based on your code quality suggestions, but it sounds like you would still close it if even if I address the issues. Is this correct?
> 
> Absolutely not! I would not close it under these assumptions
> 
> If you can copy the original PR, fix all the issues i commented under the original PR here: [#16913 (comment)](https://github.com/open-webui/open-webui/pull/16913#issuecomment-3939185585) (in case you missed it) then that would be greatly appreciated.
> 
> > I'll admit, this was the first feature I worked on out of anything and I wasn't sure that I was going to continue to use this software because it seemed to be lacking and the original PR being completely abandoned wasn't a good sign for me.
> 
> I agree. That's why I did a code review on the original PR to push things forward.
> 
> > Also, as far as the name YOLO is concerned, I'm sure the original authors intended it as an easter-egg-like joke
> 
> What? Then who suggested YOLO?

I think you misunderstand. The original PR introduces issues that I don't think would be appropriate to merge as-is. To quote myself from my reply to the discussion:

>  If I ship it as it was originally created, it would affect every tool without the ability to turn it off. It would also allow you to allow tools always by mistake without a way to revoke it.

I provided a separate comment with my explanation of the term YOLO, so I guess your question is who asked for it. It is needed as I just mentioned. It addressed my own concern as well as the concern in the discussion here: https://github.com/open-webui/open-webui/discussions/16701#discussioncomment-14154211

I really don't mind changing the name/ label though. It is quite silly and might not represent the professional vibe of the rest of the app.

**Comment by Classic298** [comment]
Ok i think i understand. So what you are proposing is:

original PR, rebased, issues fixed - plus ability to set tools to be affected by HITL or bypass HITL altogether?

For the Yolo mode, let's discuss it in discussions - and i think we can ship this later. An "allow for this chat" will suffice for the beginning i suppose - since you would be able to disable HITL altogether or for specific tools if you don't want it for now.
